src/Microsoft.Data.Analysis/DataFrame.Join.cs

```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
-                        {
-                            leftRowIndices.Append(i);
-                            rightRowIndices.Append(null);
-                        }
-                    }
-                    else
-                    {
-                        foreach (long row in otherColumnNullIndices)
-                        {
-                            leftRowIndices.Append(i);
-                            rightRowIndices.Append(row);
-                        }
-                    }
-                }
-            }
-            else if (joinAlgorithm == JoinAlgorithm.Right)
+            if (retainedJoinColumnNames.Length != supplemetaryJoinColumnNames.Length)
+                throw new ArgumentException(Strings.MismatchedArrayLengths, nameof(retainedJoinColumnNames));  
+            
+
+            HashSet<long> intersection = calculateIntersection ? new HashSet<long>() : null;
+
+            // Get occurrences of values in columns used for join in the retained and supplementary dataframes
+            Dictionary<long, ICollection<long>> occurrences = null;
+            Dictionary<long, long> retainedIndicesReverseMapping = null;
+            
+            HashSet<long> supplementaryJoinColumnsNullIndices = new HashSet<long>();
+
+            
+            for (int colNameIndex =  0; colNameIndex < retainedJoinColumnNames.Length; colNameIndex++)
             {
-                DataFrameColumn thisColumn = Columns[leftJoinColumn];
-                Dictionary<TKey, ICollection<long>> multimap = thisColumn.GroupColumnValues<TKey>(out HashSet<long> thisColumnNullIndices);
+                DataFrameColumn shrinkedRetainedColumn = retainedDataFrame.Columns[retainedJoinColumnNames[colNameIndex]];
 
-                DataFrameColumn otherColumn = other.Columns[rightJoinColumn];
-                for (long i = 0; i < otherColumn.Length; i++)
+                //shrink retained column by row occurrences from previouse step
+                if (occurrences != null)
                 {
-                    var otherColumnValue = otherColumn[i];
-                    if (otherColumnValue != null)
-                    {
-                        if (multimap.TryGetValue((TKey)otherColumnValue, out ICollection<long> rowNumbers))
-                        {
-                            foreach (long row in rowNumbers)
-                            {
-                                leftRowIndices.Append(row);
-                                rightRowIndices.Append(i);
-                            }
-                        }
-                        else
-                        {
-                            leftRowIndices.Append(null);
-                            rightRowIndices.Append(i);
-                        }
-                    }
-                    else
+                    //only rows with occurences from previose step should go for futher processing
+                    var shrinkedRetainedIndices = occurrences.Keys.ToArray();
```

Comment:
Q for next round of review: Is the call to `ToArray` necessary? I'm thinking about how performant this method might be.

Reply:
This is ok. If users are merging with multiple columns, I think it;s ok to pay this price.